### PR TITLE
Migrate recursive join to new GDS framework

### PR DESCRIPTION
**Pull request by andyfengHKU**
# Description

Please include a summary of the changes and the related issue (if applicable). Please also include
relevant motivation and context.

Fixes # (issue)

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: test/test_files/var_length_extend/n_1.test [hunk 1]
```diff
@@ -12,6 +12,9 @@
 
 -CASE VarLengthExtendN1Tests
 
+-STATEMENT CALL enable_gds = true;
```
Comment (andyfengHKU) [inline test/test_files/var_length_extend/n_1.test:15]: remove this

File: src/processor/operator/gds_call_shared_state.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            std::make_shared<processor::FactorizedTable>(mm, fTable->getTableSchema()->copy());
+        localTables.push_back(table);
+        availableLocalTables.push(table.get());
+    }
+    auto result = availableLocalTables.top();
+    availableLocalTables.pop();
+    return result;
+}
+
+void GDSCallSharedState::reclaimLocalTable(FactorizedTable* table) {
```
Comment (semihsalihoglu-uw) [inline src/processor/operator/gds_call_shared_state.cpp:29]: I think 'reclaim' is the wrong word here. I think 'return' is a better word.

File: src/processor/operator/recursive_extend/path_property_probe.cpp [hunk 3]
```diff
@@ -65,9 +65,24 @@ static void copyInternalID(ValueVector* srcVector, ValueVector* dstIDVector,
     }
 }
 
+// TODO(Xiyang): revisit me. Instead of printing in src->dst order. Maybe left->right order make
```
Comment (semihsalihoglu-uw) [inline src/processor/operator/recursive_extend/path_property_probe.cpp:68]: I would do this todo now. It will be much more difficult to come back to this to handle these edge cases.

File: test/c_api/connection_test.cpp [hunk 4]
```diff
@@ -138,52 +138,52 @@ TEST_F(CApiConnectionTest, ExecuteError) {
...
-    ASSERT_NE(result._query_result, nullptr);
-    auto resultCpp = static_cast<QueryResult*>(result._query_result);
-    ASSERT_NE(resultCpp, nullptr);
-    ASSERT_FALSE(resultCpp->isSuccess());
-    ASSERT_EQ(resultCpp->getErrorMessage(), "Interrupted.");
-    kuzu_query_result_destroy(&result);
-    t.join();
-}
-#endif
+// TEST_F(CApiConnectionTest, QueryTimeout) {
```
Comment (semihsalihoglu-uw) [inline test/c_api/connection_test.cpp:141]: What's the plan to uncomment this test? Will you merge in this commented way? If so, open an issue that will block the release. We should solve these issues before the release.

File: test/main/connection_test.cpp [hunk 5]
```diff
@@ -54,16 +54,16 @@ TEST_F(ApiTest, ParallelConnect) {
...
-    std::thread longRunningQueryThread(executeLongRunningQuery, conn.get());
-#ifdef _WIN32
-    Sleep(1000);
-#else
-    sleep(1 /* sleep 1 second before interrupt the query */);
-#endif
-    conn->interrupt();
-    longRunningQueryThread.join();
-}
+// TEST_F(ApiTest, Interrupt) {
```
Comment (semihsalihoglu-uw) [inline test/main/connection_test.cpp:57]: Ditto.

File: test/optimizer/optimizer_test.cpp [hunk 6]
```diff
@@ -104,16 +104,16 @@ TEST_F(OptimizerTest, RemoveUnnecessaryJoinTest) {
...
-    auto q1 = "MATCH (a:person)-[e:knows* SHORTEST 1..5]->(b:person) "
-              "WHERE b.ID < 0 "
-              "RETURN e, a.fName;";
-    ASSERT_STREQ(getEncodedPlan(q1).c_str(), "HJ(a._ID){RE(a)Filter()S(b)}{S(a)}");
-    auto q2 = "MATCH (a:person)-[e:knows* SHORTEST 1..5]->(b:person) "
-              "HINT (a JOIN e) JOIN b "
-              "RETURN length(e);";
-    ASSERT_STREQ(getEncodedPlan(q2).c_str(), "RE_NO_TRACK(b)S(a)");
-}
+// TEST_F(OptimizerTest, RecursiveJoinTest) {
```
Comment (semihsalihoglu-uw) [inline test/optimizer/optimizer_test.cpp:107]: Ditto.

File: test/test_files/read_list/var_length_large_adj_list_extend.test [hunk 7]
```diff
@@ -29,7 +29,7 @@
 60003
 
 # By the above formula, 28th, 29th, and 30th levels will contain 29*5K + 1, 30*5K + 1, and 31*5K+1 = 90*5K+3=450003.
--LOG KnowsVeryLargeAdjListLongPathTest
--STATEMENT MATCH (a:person)-[:knows*28..30]->(b:person) RETURN COUNT(*)
----- 1
-450003
+#-LOG KnowsVeryLargeAdjListLongPathTest
```
Comment (semihsalihoglu-uw) [inline test/test_files/read_list/var_length_large_adj_list_extend.test:32]: Ditto. For each of these commented out tests, open an issue.

File: test/test_files/var_length_extend/n_n.test [hunk 8]
```diff
@@ -40,11 +40,11 @@
 ---- 1
 1404
 
--LOG KnowsLongPathTest
--STATEMENT MATCH (a:person)-[:knows*8..11]->(b:person) RETURN COUNT(*)
--PARALLELISM 1
----- 1
-1049760
+# TODO(Xiyang): fix me
```
Comment (semihsalihoglu-uw) [inline test/test_files/var_length_extend/n_n.test:43]: Ditto.

File: src/include/processor/operator/gds_call_shared_state.h [hunk 9]
```diff
@@ -62,9 +58,18 @@ struct GDSCallSharedState {
...
     NodeOffsetMaskMap* getPathNodeMaskMap() const { return pathNodeMask.get(); }
 
+    FactorizedTable* claimLocalTable(storage::MemoryManager* mm);
+    void reclaimLocalTable(FactorizedTable* table);
+    void mergeLocalTables();
+
 private:
     std::unique_ptr<NodeOffsetMaskMap> inputNodeMask = nullptr;
     std::unique_ptr<NodeOffsetMaskMap> pathNodeMask = nullptr;
+    // We implement a local ftable pool to avoid generate many small ftables when running GDS.
```
Comment (semihsalihoglu-uw) [inline src/include/processor/operator/gds_call_shared_state.h:68]: I would remove these comments. Instead I would open an issue that explains the alternative (and I think better) solution and assign it to Guodong and Ben.

File: src/function/gds/rec_joins.cpp [hunk 10]
```diff
@@ -116,58 +118,56 @@ void SPAlgorithm::bind(const expression_vector& params, Binder* binder,
...
 void RJAlgorithm::exec(processor::ExecutionContext* executionContext) {
+    auto clientContext = executionContext->clientContext;
     auto inputNodeMaskMap = sharedState->getInputNodeMaskMap();
     for (auto& tableID : sharedState->graph->getNodeTableIDs()) {
         if (!inputNodeMaskMap->containsTableID(tableID)) {
             continue;
         }
+        auto localTable = clientContext->getTx()->getLocalStorage()->getLocalTable(tableID);
+        if (localTable && localTable->cast<storage::LocalNodeTable>().getNumRows()) {
+            throw RuntimeException("Current recursive algorithm does not work with uncommited "
```
Comment (semihsalihoglu-uw) [inline src/function/gds/rec_joins.cpp:168]: Can you open an issue about this and assign to Guodong. Let's keep track of these things.

**Review by semihsalihoglu-uw — APPROVED** [review]

File: src/processor/operator/recursive_extend/path_property_probe.cpp (hunk 3 above)
Comment (andyfengHKU) [reply src/processor/operator/recursive_extend/path_property_probe.cpp:68]: Yeah I'll do this before the release.

File: test/optimizer/optimizer_test.cpp (hunk 6 above)
Comment (andyfengHKU) [reply test/optimizer/optimizer_test.cpp:107]: Removed

File: test/test_files/read_list/var_length_large_adj_list_extend.test (hunk 7 above)
Comment (andyfengHKU) [reply test/test_files/read_list/var_length_large_adj_list_extend.test:32]: Done

File: test/test_files/var_length_extend/n_n.test (hunk 8 above)
Comment (andyfengHKU) [reply test/test_files/var_length_extend/n_n.test:43]: Done

File: src/processor/operator/gds_call_shared_state.cpp (hunk 2 above)
Comment (andyfengHKU) [reply src/processor/operator/gds_call_shared_state.cpp]: Done

File: test/c_api/connection_test.cpp (hunk 4 above)
Comment (andyfengHKU) [reply test/c_api/connection_test.cpp]: Uncommented.

File: test/main/connection_test.cpp (hunk 5 above)
Comment (andyfengHKU) [reply test/main/connection_test.cpp]: Uncommented.

File: src/include/processor/operator/gds_call_shared_state.h (hunk 9 above)
Comment (andyfengHKU) [reply src/include/processor/operator/gds_call_shared_state.h:68]: Done
